Add how to add a recipe documentation (#3614)
ESMValGroup/ESMValTool: esmvaltool/utils/recipe_test_workflow/doc/source/user_guide/adding_a_recipe.rst

Adding a recipe to the Recipe Test Workflow (|RTW|)
===================================================

.. include:: ../common.txt

**Please note:** Before you follow these steps to add your recipe, you must be

> able to successfully run it with ESMValtool without error.

Author: Added in d3f1c3fe5

* Run ESMValTool locally with your recipe, make sure to take note of the memory
  and time expenditure provided by the terminal output::

    esmvaltool run <your_recipe_name.yml>

> 

Reviewer: This is available in the [latest version](https://github.com/ESMValGroup/ESMValTool/blob/d236a324278f75c2cbb41d133e88f16dd18c72f7/esmvaltool/utils/recipe_test_workflow/doc/source/user_guide/adding_a_recipe.rst?plain=1#L23C3-L26C1), so I will resolve this on @chrisbillowsMO's behalf, as discussed 😊

* Run the same recipe on JASMIN, again taking note of the memory and time
  expenditure printed to the terminal.

* Stop any running recipe_test_workflow workflows::

    cylc stop "a_running_recipe_test_workflow"

* Add the recipe to the ``[task parameters]`` section of the |RTW| ``flow.cylc``
  workflow file (make sure to place it within the "fast" or "medium" category

>   depending on how long it took to run):

Reviewer 2: Formatting the last row like this `recipe_new_recipe \,` seems to throw an error.

```bash
IllegalValueError: (type=parameter) [task parameters]fast = recipe_radiation_budget,         recipe_albedolandcover,         recipe_heatwaves_coldwaves,         recipe_autoassess_landsurface_soilmoisture,         recipe_consecdrydays,         examples--recipe_python,         recipe_ocean_multimap,     medium = recipe_ensclus  # Should be in fast: is in medium temporarily to - (medium = recipe_ensclus: bad value)

```

>   depending on how long it took to run):

Reviewer: > > * we also discussed adding the "slow" recipe option and criteria in the docs
> > * and adding a slow recipe to the `flow.cylc` so the section exists and isn't empty, to help guide a user placing a recipe there
> 
> This will be added once Emma has picked a slow recipe to use

I'm going to suggest this work is done via #3570.

> * Run the workflow. The process task for the new recipe should succeed, but its

Reviewer 2: My recipe process task failed with this error: `slurmstepd: error: Exceeded step memory limit at some point.`

When run locally the recipe required:

```
2024-05-28 08:09:40,810 UTC [127076] INFO    Time for running the recipe was: 0:06:52.260589
2024-05-28 08:09:41,170 UTC [127076] INFO    Maximum memory used (estimate): 2.5 GB
```

Reviewer: The latest instructions require that the duration and memory usage are taken from a run [on the compute server you use at your site](https://github.com/ESMValGroup/ESMValTool/blob/d236a324278f75c2cbb41d133e88f16dd18c72f7/esmvaltool/utils/recipe_test_workflow/doc/source/user_guide/adding_a_recipe.rst?plain=1#L6C1-L11C1), so the process task should not fail for this reason 👍 I will resolve this on @chrisbillowsMO's behalf, as discussed 😊

  compare task should fail with an error that the reference data (KGO) does
  not exist::

    cylc vip -O <your_site_name>

* Locate the workflow run folder of the workflow you just completed.

> * Copy the output files from the "cycle" folder (run/share/bin/cycle) of the

Reviewer 2: I think this should be run/share/cycle... bin only contains the `rtw-env` file?

  workflow you just ran, to your site specific KGO rootpath folder
  (this folder should be set as the value for the variable "KGO_ROOT_PATH="

>   in your rose-suite<your_site>.conf file)::

Reviewer 2: is it worth specifying where this file is?

    cp -r <directory_of_recipe_output_cycle_folder> <KGO_rootpath_folder>

* Change directory to the rootpath KGO directory::

    cd <KGO_rootpath_directory>

* Allow write permissions for all users on the directory and it's
  subdirectories of the recipe you've added to the KGO folder::

    chmod a+w <the_directory_of_the_recipe_you_have_copied_into_the_KGO_folder>

* Run the RTW again::

    cylc vip -O <your_site_name>

* The workflow should now succeed.

* Take note of how long the run took  to complete on cylc. This can be found in the `job.time` section of the task listed as `process_<your_recipe>`.

> * Locate your local <site>.cylc file, found in (recipe_test_workflow/site).

Reviewer 2: So I've followed steps from here to fix the issue I got with the very first process task run.

Reviewer: @chrisbillowsMO, can this comment be resolved? Is there anything needed here? 🤔

Reviewer 2: No - it was just a note on the path/order I ended up following through the instructions as they were then. I'm sure that's moot now so I'll resolve.

* Open it in your preferred code editor.

* Locate the `COMPUTE` section, it should look something like this::

    [[COMPUTE]]
    platform = <your_platform_here>
    execution time limit = PT2M
    [[[directives]]]
    --wckey = RTW
    --ntasks = {{ MAX_PARALLEL_TASKS }}
    --mem = 2G

* Compare the `execution time limit` and --mem (memory) units here with the
  readings you took locally. If your local readings do not exceed these, then
  you have successfully added the recipe to the workflow, and can now commit
  and push your changes.

* If either of the time/memory readings from your local run are larger than the
  values specified in the `COMPUTE` section, you need to add your recipe as
  another `process` similar to::

    [[process<fast=recipe_albedolandcover>]]
    # Actual: 0m31s, 2.5 GB on 2024-04-08.
    execution time limit = PT2M
    [[[directives]]]
        --mem = 3G

> * Variable (fast, medium) must be consistent with flow.cylc.

Reviewer 2: If this is now moved into the documentation then that should be specified here.

Reviewer: I have re-raised this in https://github.com/ESMValGroup/ESMValTool/pull/3614#discussion_r1644087547 👍 I will resolve this on @chrisbillowsMO's behalf, as discussed 😊

* The commented "Actual" reading should be the time/memory
  reading you recorded from cylc review.

* Adjust the values for `execution_time` and `--mem` to be larger than the
  values you recorded from cylc review.

* Stop any running workflows.

* Run the recipe test workflow again.

* If the workflow succeeds then your recipe has successfully been added to the

>   workflow. You can now commit your changes and push them onto GitHub.

Reviewer 2: Add an instruction to add the recipe to docs in `tested_recipes.rst`.  The recipes should be listed alphabetically (perhaps specify how to order recipes in sub-directories?)

> Adding a recipe to the Recipe Test Workflow (|RTW|)

Reviewer: Would it be possible for the title to be "How to add a recipe to the Recipe Test Workflow (|RTW|)", please? 😊

Author: Amended in commit d01a94f42

===================================================

.. include:: ../common.txt

**Please note**: Before you follow these steps to add it, you must be

> able to successfully run it with the latest version of ESMValTool on the

Author: Amended in commit 7e4c9855a

compute server you use at your site, as detailed by the ``platform`` option in
the ``[[COMPUTE]]`` section in the site-specific ``.cylc`` file in the
``esmvaltool/utils/recipe_test_workflow/recipe_test_workflow/site/`` directory.

> 

Author: Added in commit d29e76c62

#. Obtain the duration and memory usage of the recipe from the messages printed

>    to screen after running your recipe on the compute cluster you use at your

Author: Corrected in commit 8c9f0e1dd

   site; these messages will look something like::

    YYYY-MM-DD HH:MM:SS:sss UTC [12345] INFO    Time for running the recipe was: 0:02:13.334742
    YYYY-MM-DD HH:MM:SS:sss UTC [12345] INFO    Maximum memory used (estimate): 2.4 GB
    [...]
    YYYY-MM-DD HH:MM:SS:sss UTC [12345] INFO    Run was successful

#. Add the recipe to the ``[task parameters]`` section in the
   ``esmvaltool/utils/recipe_test_workflow/recipe_test_workflow/flow.cylc``
   file. If the recipe takes less than 10 minutes to run then it should be
   added to the ``fast`` option. Recipes that take longer than ten minutes

>    should be added to the ``medium`` option.

Reviewer: We should extend this to mention the syntax that @chrisbillowsMO raised at https://github.com/ESMValGroup/ESMValTool/pull/3614#discussion_r1617145289.

Author: I've had a go at describing the syntax in commit 5569fb08e

#. If the memory usage of the recipe is larger than the value specified by the
   ``--mem`` option in the ``[[[directives]]]`` section in the ``[[COMPUTE]]``
   section in the aforementioned site-specific ``.cylc`` file, add a section
   (in alphabetical order) to this file as shown below (make sure to round off

>    the duration to the nearest second when you add it)::

Reviewer: I just realised we should add something here about the values that are used below. The guidance we are provided at the MO states:

> Advice on estimating appropriate resource limits will vary from job to job, but we suggest setting a wallclock limit of, say, 10-20% more than the maximum expected/recorded run length, and try not to regularly waste more than e.g. 500 MiB per core in memory estimation.

We should say something like this in our instructions. Perhaps also state that any actual duration less than ~1m45s should use an `execution time limit` of `PT2M`, and perhaps round up to the nearest integer for the memory usage (I am sure (from experience, I can't see anything in the [documentation](https://slurm.schedmd.com/sbatch.html#OPT_mem)) that the value for `--mem` can only be an integer).

>    the duration to the nearest second when you add it)::

Reviewer: Also, it should be highlighted that the `fast` in the task definition below must correspond to the option where the recipe was added in the `flow.cylc` file, as @chrisbillowsMO mentioned at https://github.com/ESMValGroup/ESMValTool/pull/3614#discussion_r1617128306.

>    the duration to the nearest second when you add it)::

Author: I've made an attempt at adjusting this according to the two comments above in commit 2be0d9c76

>    the duration to the nearest second when you add it)::

Reviewer: Thank you! I have made some further suggestion on this in my most recent review, so will resolve this now 👍

    [[process<fast=recipe_albedolandcover>]]
    # Actual: 0m31s, 2.5 GB on 2024-04-08.
    execution time limit = PT2M
    [[[directives]]]
        --mem = 3G

#. Stop any running ``recipe_test_workflow`` workflows::

    cylc stop recipe_test_workflow

#. Run the recipe test workflow (it is expected that the ``compare`` task will

>    fail).

Author: Added in commit ac585d2c2

> #. Update the KGO:

Author: I've now corrected this and added a section in the glossary for KGO's in commit d9c4fc2e8

        #. Locate the workflow run folder of the workflow you just completed.

        #. Recursively copy the recipe output directory (i.e.
           ``recipe_<recipe>_<date>_<time>``) from the
           ``${HOME}/cylc-run/recipe_test_workflow/run1/share/cycle/<cycle>``
           directory to your site-specific KGO directory, as detailed by the
           ``KGO_ROOT_PATH`` option in the site-specific ``.conf`` file in the
           ``esmvaltool/utils/recipe_test_workflow/recipe_test_workflow/opt/``
           directory::

            cp -r <directory_of_recipe_output_cycle_folder> <KGO_rootpath_folder>

        #. Change directory to the rootpath KGO directory::

            cd <KGO_rootpath_directory>

        #. Allow write permissions for all users on the directory and it's
           subdirectories of the recipe you've added to the KGO folder.::

>             chmod a+w <the_directory_of_the_recipe_you_have_copied_into_the_KGO_folder>

Author: This should now be corrected in commit a648f6abe

>             chmod a+w <the_directory_of_the_recipe_you_have_copied_into_the_KGO_folder>

Reviewer: The points still show as quoted for me 🙁

>             chmod a+w <the_directory_of_the_recipe_you_have_copied_into_the_KGO_folder>

Author: I'm hoping this has now been fixed in commit 0c4c28446

>             chmod a+w <the_directory_of_the_recipe_you_have_copied_into_the_KGO_folder>

Reviewer: Looks good, thank you! 😁

#. Stop any running ``recipe_test_workflow`` workflows::

    cylc stop recipe_test_workflow

> #. Run the RTW again. The workflow should now succeed.::

Author: Corrected in commit cd4d18714

Adding a recipe to the Recipe Test Workflow (|RTW|)
===================================================

.. include:: ../common.txt

**Please note**: Before you follow these steps to add your recipe, you must be
able to successfully run it with the latest version of ESMValTool on the
compute server you use at your site (you should be able to find this in your
site specific ``.cylc`` file in the site directory).

#. Stop any running recipe_test_workflow workflows::

>     cylc stop "a_running_recipe_test_workflow"

Reviewer: Also `cylc stop recipe_test_workflow` is the command everyone should use 👍

#. Locate your local ``<site>.cylc`` file, found in (recipe_test_workflow/site)

>    .

Reviewer: Weird, this came up twice! You addressed it here: https://github.com/ESMValGroup/ESMValTool/pull/3614#discussion_r1619045568 😊

#. Open it in your preferred code editor.

#. Locate the ``[[COMPUTE]]`` section, it should look something like this::

    [[COMPUTE]]
    platform = <your_platform_here>
    execution time limit = PT2M
    [[[directives]]]
    --wckey = RTW
    --ntasks = {{ MAX_PARALLEL_TASKS }}

>     --mem = 2G

Reviewer: Weird, this one came up twice too! 😮

#. Run your recipe with ESMValTool on your compute server for your site.

#. If the recipe takes less than 10 minutes to run then it should be added as a
   "fast" recipe in the ``flow.cylc`` file within the ``[task parameters]``
   section.If it takes longer than ten minutes it should be included in "medium
   ".

#. If either of the memory readings from your run are larger than the
   values specified in the ``[[COMPUTE]]`` section, you need to add your recipe as
   another `process` similar to::

    [[process<fast=recipe_albedolandcover>]]
    # Actual: 0m31s, 2.5 GB on 2024-04-08.
    execution time limit = PT2M
    [[[directives]]]
        --mem = 3G

> 

Reviewer: Add a note here to say to round to the nearest second in the comment (i.e. there is no need to copy the full duration from the messages printed to screen), see https://github.com/ESMValGroup/ESMValTool/pull/3627#discussion_r1618743351.

#. Run the recipe test workflow.

#. The ``compare`` task will fail. These next steps should fix it,

>    and even if the ``compare`` task has passed, they must still be completed.

Reviewer: I can't think of a case, when adding a new recipe, that the ``compare`` task will pass, so this point can be removed.

#. Locate the workflow run folder of the workflow you just completed.

> #. Copy the output files from the "cycle" folder (``run/share/cycle``) of the
>    workflow you just ran, to your site specific KGO rootpath folder
>    (this folder should be set as the value for the variable ``KGO_ROOT_PATH=``
>    in your ``rose-suite<your_site>.conf`` file found in the ``/data/users/esmva
>    l/KGO`` directory)::

Reviewer: ```suggestion
#. Recursively copy the recipe output directory (i.e.
   ``recipe_<recipe>_<date>_<time>``) from the
   ``${HOME}/cylc-run/recipe_test_workflow/run1/share/cycle/<cycle>`` directory
   to your site-specific KGO directory, as detailed by the ``KGO_ROOT_PATH``
   option in the site-specific ``.conf`` file in the
   ``esmvaltool/utils/recipe_test_workflow/recipe_test_workflow/opt/``
   directory::
```

    cp -r <directory_of_recipe_output_cycle_folder> <KGO_rootpath_folder>

#. Change directory to the rootpath KGO directory::

    cd <KGO_rootpath_directory>

#. Allow write permissions for all users on the directory and it's
   subdirectories of the recipe you've added to the KGO folder::

>     chmod a+w <the_directory_of_the_recipe_you_have_copied_into_the_KGO_folder>

Reviewer: Would it be possibe to create an "Update the KGO" instruction that have these points as sub-points, please?

#. Stop any running recipe_test_workflow workflows::

    cylc stop "a_running_recipe_test_workflow"

#. Run the RTW again::

>     cylc vip -O <your_site_name>

Reviewer: ```suggestion
    cylc vip -O <site>
```

> #. The workflow should now succeed.

Reviewer: Add this text to the previous instruction (it's not an instruction on its own).

#. Add your recipe to the list of "Currently tested recipes" in ``tested_recipes.rst``
   (this file is located within ``recipe_test_workflow/doc/source``) in
   it's correct position (the list is ordered alphabetically), this will add
   it to the documentation.

#. If the workflow succeeds then your recipe has successfully been added to the

>    workflow. You can now commit your changes and push them onto GitHub.

Reviewer: ```suggestion
#. Commit and push your changes.

Adding a recipe to the Recipe Test Workflow (|RTW|)
===================================================

.. include:: ../common.txt

**Please note:** Before you follow these steps to add your recipe, you must be
able to successfully run it with ESMValtool without error.

* Run ESMValTool locally with your recipe, make sure to take note of the memory
  and time expenditure provided by the terminal output::

    esmvaltool run <your_recipe_name.yml>

* Run the same recipe on JASMIN, again taking note of the memory and time
  expenditure printed to the terminal.

> * Stop any running workflows::

Author: Thanks Alistair! I've attempted a change for this in commit 06f79f23d

    cylc stop "*"

* Add the recipe to the ``[task parameters]`` section of the |RTW| ``flow.cylc``
  workflow file (make sure to place it within the "fast" or "medium" category
  depending on how long it took to run):

> * Run the workflow (it should fail)::

Author: Fixed in commit 06f79f23d

    cylc vip -O <your_site_name>

* Locate the workflow run folder of the workflow you just completed.

* Copy the output files from the "cycle" folder (run/share/bin/cycle) of the
  workflow you just ran, to your site specific KGO rootpath folder

>   (/data/users/esmval/KGO)::

Reviewer 3: KGO rootpath is machine-specific, so please refer to the file/variable where it is set, rather than giving this site-specific example.

Author: I've had a go at a change for this in commit 06f79f23d
